src/python/pants/core/goals/check.py

```diff
@@ -201,8 +203,33 @@ async def check(
         )
         for request_type in request_types
     )
+
+    environment_names = await MultiGet(
+        Get(
+            EnvironmentName,
+            EnvironmentNameRequest,
+            EnvironmentNameRequest.from_field_set(field_set),
+        )
+        for request in requests
+        for field_set in request.field_sets
+    )
+    request_for_each_environment_name = (
+        # this is actually correct:
+        request
+        for request in requests
+        for _ in request.field_sets
+    )
+    grouped = itertools.groupby(
+        zip(request_for_each_environment_name, environment_names), lambda x: x[0]
+    )
+    exploded_requests = [
+        (request, {env_name for (_, env_name) in env_names}) for request, env_names in grouped
+    ]
```

Comment:
It's not clear what this is doing... but it looks a bit like the `for-for` sortof undoes some of the grouping here, so you could maybe skip "exploding"?